pennylane/argmap.py

```diff
@@ -0,0 +1,251 @@
+    this circuit, depending on the parameter that controls them:
+
+    .. code-block:: python
+
+        generators = qml.ArgMap({
+            (0, 0): qml.PauliZ(0),
+            (0, 1): qml.PauliY(0),
+            (0, 2): qml.PauliZ(0),
+            1: qml.PauliX(0),
+            (1, 0): qml.PauliZ(1),
+            (1, 1): qml.PauliY(1),
+            (1, 2): qml.PauliZ(1),
+        })
+
+    Then the generators can be accessed by using the tuples ``(arg_index, par_index)`` as key
+    with ``par_index=None`` for the ``PauliX(0)`` generator. The latter may alternatively
+    be accessed directly via ``generators[1]``.
+    """
+
+    def __init__(self, data=None, single_arg=False, single_object=False, like=None):
+        if like is not None:
+            if not isinstance(like, ArgMap):
+                raise ArgMapError("Trying to inherit properties from non-ArgMap object.")
+            single_arg, single_object = like.single_arg, like.single_object
+        self.single_arg = single_arg
+        self.single_object = single_object
+        _data = self._preprocess_data(data)
+        super().__init__(_data)
+        self.consistency_check()
+
+    def _preprocess_data(self, data):
+        if data is None:
+            data = {}
+        if self.single_object:
+            return {(None, None): data}
+
+        if not isinstance(data, dict):
+            try:
+                data = dict(data)
+            except (ValueError, TypeError) as e:
+                raise ArgMapError(
+                    f"The input could not be interpreted as dictionary; input:\n{data}"
+                ) from e
+
+        return {_interpret_key(key, self.single_arg): val for key, val in data.items()}
+
+    def __getitem__(self, key):
+        key = _interpret_key(key, self.single_arg)
+        return super().__getitem__(key)
+
+    def __setitem__(self, key, value):
+        key = _interpret_key(key, self.single_arg)
+        return super().__setitem__(key, value)
+
+    def setdefault(self, key, default=None):
+        key = _interpret_key(key, self.single_arg)
+        return super().setdefault(key, default)
+
+    def __eq__(self, other):
+        if isinstance(other, ArgMap):
+            if (self.single_arg + other.single_arg) % 2:
+                return False
+            if (self.single_object + other.single_object) % 2:
+                return False
+        return super().__eq__(other)
+
+    def __ne__(self, other):
+        if isinstance(other, ArgMap):
+            if (self.single_arg + other.single_arg) % 2:
+                return True
+            if (self.single_object + other.single_object) % 2:
+                return True
+        return super().__ne__(other)
+
+    def get(self, key, default=None):
+        key = _interpret_key(key, self.single_arg)
+        return super().get(key, default)
+
+    def consistency_check(self, check_values=False):
+        r"""Check the stored keys and optionally values for consistency."""
```

Comment:
What does `check_values` do? Since this is a public method, make sure to document the behaviour, arguments, and return type!

Reply:
(an example would help as well :) )

Reply:
:sweat_smile: